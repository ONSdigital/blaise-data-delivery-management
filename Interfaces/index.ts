interface DataDeliveryBatchData {
    survey?: string
    date: Date
    dateString: string
    name: string
}

interface DataDeliveryFile {
    prefix: string
    instrumentName: string
}

interface DataDeliveryFileStatus {
    batch: string
    dd_filename: string
    state: string
    updated_at: string
    instrumentName: string
}


export type {DataDeliveryFile, DataDeliveryBatchData, DataDeliveryFileStatus};
