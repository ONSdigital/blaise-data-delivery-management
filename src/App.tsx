import React, { ReactElement, useEffect, useState } from "react";
import { Switch, Route, useLocation } from "react-router-dom";
import {
    Footer,
    Header,
    ONSPanel,
    DefaultErrorBoundary,
    ErrorBoundary,
} from "blaise-design-system-react-components";
import Confirmation from "./Components/Confirmation";
import BatchesList from "./Components/BatchesList";
import BatchStatusList from "./Components/BatchStatusList";
import { getBatchStatusDescriptions } from "./utilities/http";
import "./style.css";

const divStyle = {
    minHeight: "calc(67vh)"
};

interface Location {
    state: { status: string }
}

type BatchDescription = { [key: string]: string }

function App(): ReactElement {

    const location = useLocation();
    const { status } = (location as Location).state || { status: "" };
    const [statusDescriptionList, setStatusDescriptionList] = useState<BatchDescription>({});

    useEffect(() => {
        callGetBatchStatusDescriptions().then(() => console.log("getBatchStatusDescriptions Complete"));
    }, []);

    async function callGetBatchStatusDescriptions() {
        setStatusDescriptionList({});

        const [success, statusDescriptionList] = await getBatchStatusDescriptions();

        if (!success) {
            return;
        }

        setStatusDescriptionList(statusDescriptionList as BatchDescription);
    }

    return (
        <>
            <Header title={"Data Delivery Management"} />
            <div style={divStyle} className="ons-page__container ons-container">
                <DefaultErrorBoundary>
                    <Switch>
                        <Route path="/trigger">
                            <Confirmation />
                        </Route>
                        <Route path="/batch">
                            <BatchStatusList statusDescriptionList={statusDescriptionList} />
                        </Route>
                        <Route path="/">

                            {
                                status !== "" &&
                                <ONSPanel status={status?.includes("success") ? "success" : "error"}>
                                    <p>{status}</p>
                                </ONSPanel>
                            }

                            <main id="main-content" className="ons-page__main ons-u-mt-no">

                                {/*<ul className="list list--bare list--inline u-mt-m">*/}
                                {/*    <li className="list__item">*/}
                                {/*        <Link to="/trigger" id="audit-logs-link">*/}
                                {/*            Trigger Data Delivery*/}
                                {/*        </Link>*/}
                                {/*    </li>*/}
                                {/*</ul>*/}

                                <h1 className="ons-u-mt-m">Data delivery runs</h1>
                                <ErrorBoundary errorMessageText={"Unable to load batch list table correctly"}>
                                    <BatchesList />
                                </ErrorBoundary>
                            </main>
                        </Route>
                    </Switch>
                </DefaultErrorBoundary>
            </div>
            <Footer />
        </>
    );
}

export default App;
