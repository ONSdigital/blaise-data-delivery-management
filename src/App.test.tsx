import React from "react";
import {render, waitFor, cleanup} from "@testing-library/react";
import App from "./App";
import "@testing-library/jest-dom";
import flushPromises, {mock_server_request_Return_JSON} from "./tests/utils";
import {act} from "react-dom/test-utils";
import {createMemoryHistory} from "history";
import {Router} from "react-router";
import {DataDeliveryBatchData} from "../Interfaces";

describe("React homepage", () => {

    const batches: DataDeliveryBatchData[] = [
        {
            survey: "OPN",
            date: new Date("2021-03-24T11:30:00.000Z"),
            dateString: "24/03/2021 11:30:00",
            name: "OPN_24032021_113000"
        },
        {
            survey: "OPN",
            date: new Date("2021-03-12T02:30:00.000Z"),
            dateString: "12/03/2021 02:30:00",
            name: "OPN_12032021_023000"
        }
    ];


    beforeAll(() => {
        mock_server_request_Return_JSON(200, batches);
    });

    it("view instrument page matches Snapshot", async () => {
        const history = createMemoryHistory();
        const wrapper = render(
            <Router history={history}>
                <App/>
            </Router>
        );

        await act(async () => {
            await flushPromises();
        });

        await waitFor(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it("should render correctly", async () => {
        const history = createMemoryHistory();
        const {getByText, queryByText} = render(
            <Router history={history}>
                <App/>
            </Router>
        );

        expect(queryByText(/Loading/i)).toBeInTheDocument();

        await waitFor(() => {
            expect(getByText(/Data Delivery Management/i)).toBeDefined();
            expect(queryByText(/Loading/i)).not.toBeInTheDocument();
        });

        await waitFor(() => {
            expect(getByText(/Data Delivery Management/i)).toBeDefined();
            expect(getByText(/24\/03\/2021 11:30:00/i)).toBeDefined();
            expect(getByText(/12\/03\/2021 02:30:00/i)).toBeDefined();
            expect(queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});


describe("Given the API returns an empty list", () => {

    beforeAll(() => {
        mock_server_request_Return_JSON(200, []);
    });

    it("it should render with a message to inform the user in the list", async () => {
        const history = createMemoryHistory();
        const {getByText, queryByText} = render(
            <Router history={history}>
                <App/>
            </Router>
        );

        expect(queryByText(/Loading/i)).toBeInTheDocument();


        await waitFor(() => {
            expect(getByText(/No data delivery runs found./i)).toBeDefined();
            expect(queryByText(/Loading/i)).not.toBeInTheDocument();
        });

    });

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});
