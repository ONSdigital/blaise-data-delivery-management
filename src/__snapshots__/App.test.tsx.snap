// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React homepage view instrument page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="phase-banner"
      >
        <div
          class="container "
        >
          <div
            class="grid grid--flex grid--gutterless grid--vertical-center grid--no-wrap"
          >
            <div
              class="grid__col col-auto u-flex-no-grow"
            >
              <h3
                class="phase-banner__badge"
              >
                BETA
              </h3>
            </div>
            <div
              class="grid__col col-auto u-flex-shrink"
            >
              <p
                class="phase-banner__desc u-fs-s u-mb-no"
              >
                This is a new service – your 
                <a
                  href="https://ons.service-now.com/"
                >
                  feedback
                </a>
                 will help us improve it.
              </p>
            </div>
          </div>
        </div>
      </div>
      <header
        class="header header--internal"
      >
        <div
          class="header__top"
          role="banner"
        >
          <div
            class="container"
          >
            <div
              class="header__grid-top grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap "
            >
              <div
                class="grid__col col-auto"
              >
                <a
                  class="header__logo-link"
                  href="/"
                >
                  <picture>
                    <img
                      alt="Office for National Statistics logo"
                      class="header__logo"
                      src="https://ons-design-system.netlify.app/img/ons-logo-neg-en.svg"
                    />
                  </picture>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="header__main"
        >
          <div
            class="container"
          >
            <div
              class="grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap"
            >
              <div
                class="grid__col col-auto u-flex-shrink"
              >
                <div
                  class="header__title"
                >
                  Data Delivery Management
                </div>
              </div>
            </div>
          </div>
        </div>
      </header>
      <div
        class="page__container container"
        style="min-height: calc(67vh);"
      >
        <main
          class="page__main"
          id="main-content"
        >
          <ul
            class="list list--bare list--inline u-mt-m"
          >
            <li
              class="list__item"
            >
              <a
                href="/trigger"
                id="audit-logs-link"
              >
                Trigger Data Delivery
              </a>
            </li>
          </ul>
          <h1
            class="u-mt-m"
          >
            Data delivery runs
          </h1>
          <button
            class="btn btn--small "
            data-testid="button"
            type="button"
          >
            <span
              class="btn__inner"
            >
              Reload
            </span>
          </button>
          <table
            class="table "
            id="batches-table"
          >
            <thead
              class="table__head u-mt-m"
            >
              <tr
                class="table__row"
              >
                <th
                  class="table__header "
                  scope="col"
                >
                  <span>
                    Survey
                  </span>
                </th>
                <th
                  class="table__header "
                  scope="col"
                >
                  <span>
                    Data delivery run time
                  </span>
                </th>
                <th
                  class="table__header "
                  scope="col"
                >
                  <span>
                    View run status
                  </span>
                </th>
              </tr>
            </thead>
            <tbody
              class="table__body"
            >
              <tr
                class="table__row"
                data-testid="batches-table-row"
              >
                <td
                  class="table__cell "
                >
                  OPN
                </td>
                <td
                  class="table__cell "
                >
                  24/03/2021 11:30:00
                </td>
                <td
                  class="table__cell "
                >
                  <a
                    aria-label="View run status 24/03/2021 11:30:00"
                    href="/batch/24032021_113000"
                  >
                    View run status
                  </a>
                </td>
              </tr>
              <tr
                class="table__row"
                data-testid="batches-table-row"
              >
                <td
                  class="table__cell "
                >
                  OPN
                </td>
                <td
                  class="table__cell "
                >
                  12/03/2021 02:30:00
                </td>
                <td
                  class="table__cell "
                >
                  <a
                    aria-label="View run status 12/03/2021 02:30:00"
                    href="/batch/12032021_023000"
                  >
                    View run status
                  </a>
                </td>
              </tr>
            </tbody>
          </table>
        </main>
      </div>
      <footer
        class="footer"
      >
        <div
          class="footer__body"
          data-analytics="footer"
        >
          <div
            class="container "
          >
            <div
              class="grid grid--flex grid--between"
            >
              <div
                class="grid__col"
              >
                <div
                  class="u-mt-m u-mb-s"
                >
                  <img
                    alt="Office for National Statistics"
                    src="https://ons-design-system.netlify.app/img/ons-logo-black-en.svg"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </footer>
    </div>
  </body>,
  "container": <div>
    <div
      class="phase-banner"
    >
      <div
        class="container "
      >
        <div
          class="grid grid--flex grid--gutterless grid--vertical-center grid--no-wrap"
        >
          <div
            class="grid__col col-auto u-flex-no-grow"
          >
            <h3
              class="phase-banner__badge"
            >
              BETA
            </h3>
          </div>
          <div
            class="grid__col col-auto u-flex-shrink"
          >
            <p
              class="phase-banner__desc u-fs-s u-mb-no"
            >
              This is a new service – your 
              <a
                href="https://ons.service-now.com/"
              >
                feedback
              </a>
               will help us improve it.
            </p>
          </div>
        </div>
      </div>
    </div>
    <header
      class="header header--internal"
    >
      <div
        class="header__top"
        role="banner"
      >
        <div
          class="container"
        >
          <div
            class="header__grid-top grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap "
          >
            <div
              class="grid__col col-auto"
            >
              <a
                class="header__logo-link"
                href="/"
              >
                <picture>
                  <img
                    alt="Office for National Statistics logo"
                    class="header__logo"
                    src="https://ons-design-system.netlify.app/img/ons-logo-neg-en.svg"
                  />
                </picture>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="header__main"
      >
        <div
          class="container"
        >
          <div
            class="grid grid--gutterless grid--flex grid--between grid--vertical-center grid--no-wrap"
          >
            <div
              class="grid__col col-auto u-flex-shrink"
            >
              <div
                class="header__title"
              >
                Data Delivery Management
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      class="page__container container"
      style="min-height: calc(67vh);"
    >
      <main
        class="page__main"
        id="main-content"
      >
        <ul
          class="list list--bare list--inline u-mt-m"
        >
          <li
            class="list__item"
          >
            <a
              href="/trigger"
              id="audit-logs-link"
            >
              Trigger Data Delivery
            </a>
          </li>
        </ul>
        <h1
          class="u-mt-m"
        >
          Data delivery runs
        </h1>
        <button
          class="btn btn--small "
          data-testid="button"
          type="button"
        >
          <span
            class="btn__inner"
          >
            Reload
          </span>
        </button>
        <table
          class="table "
          id="batches-table"
        >
          <thead
            class="table__head u-mt-m"
          >
            <tr
              class="table__row"
            >
              <th
                class="table__header "
                scope="col"
              >
                <span>
                  Survey
                </span>
              </th>
              <th
                class="table__header "
                scope="col"
              >
                <span>
                  Data delivery run time
                </span>
              </th>
              <th
                class="table__header "
                scope="col"
              >
                <span>
                  View run status
                </span>
              </th>
            </tr>
          </thead>
          <tbody
            class="table__body"
          >
            <tr
              class="table__row"
              data-testid="batches-table-row"
            >
              <td
                class="table__cell "
              >
                OPN
              </td>
              <td
                class="table__cell "
              >
                24/03/2021 11:30:00
              </td>
              <td
                class="table__cell "
              >
                <a
                  aria-label="View run status 24/03/2021 11:30:00"
                  href="/batch/24032021_113000"
                >
                  View run status
                </a>
              </td>
            </tr>
            <tr
              class="table__row"
              data-testid="batches-table-row"
            >
              <td
                class="table__cell "
              >
                OPN
              </td>
              <td
                class="table__cell "
              >
                12/03/2021 02:30:00
              </td>
              <td
                class="table__cell "
              >
                <a
                  aria-label="View run status 12/03/2021 02:30:00"
                  href="/batch/12032021_023000"
                >
                  View run status
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </main>
    </div>
    <footer
      class="footer"
    >
      <div
        class="footer__body"
        data-analytics="footer"
      >
        <div
          class="container "
        >
          <div
            class="grid grid--flex grid--between"
          >
            <div
              class="grid__col"
            >
              <div
                class="u-mt-m u-mb-s"
              >
                <img
                  alt="Office for National Statistics"
                  src="https://ons-design-system.netlify.app/img/ons-logo-black-en.svg"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
